test/SILOptimizer/missing_returns.swift

```diff
@@ -0,0 +1,135 @@
+
+  static subscript<A>(root: MyEmptyType) -> A {}
+
+  subscript(_ e: MyEmptyType) -> Int {}
+  subscript<T>(_ e: MyEmptyType) -> T {}
+  subscript(_ i: Int) -> Int {} // expected-error{{missing return in getter expected to return 'Int'}}
+  subscript<T>(_ p: Int) -> T {} // expected-error{{missing return in getter expected to return 'T'}}
+  subscript(_ i: Int) -> Self {} // expected-error{{getter with uninhabited return type 'MyEmptyType' is missing call to another never-returning function on all paths}}
+  subscript(_ s: Self) -> Self {}
+
+  static func unreachable_static_implicit_return(_ e: MyEmptyType) -> Int {}
+  func unreachable(_ e: MyEmptyType) -> Int { // expected-note{{'e' is of type 'MyEmptyType' which cannot be constructed because it is an enum with no cases}}
+    42 // expected-warning{{will never be executed}}
+  }
+
+  // FIXME: should these produce warnings since they implicity take an uninhabited 'self' param?
+  func implicitly_unreachable() { _ = 42 }
+  func implicitly_unreachable_implicit_return() -> Int { 42 }
+}
+
+extension Never {
+  var i: Int {} // expected-error{{missing return in getter expected to return 'Int'}}
+  var n: Never {} // expected-error{{getter with uninhabited return type 'Never' is missing call to another never-returning function on all paths}}
+
+  static subscript<A>(root: Never) -> A {}
+
+  subscript(_ n: Never) -> Int {}
+  subscript<T>(_ e: Never) -> T {}
+  subscript(_ i: Int) -> Int {} // expected-error{{missing return in getter expected to return 'Int'}}
+  subscript<T>(_ p: Int) -> T {} // expected-error{{missing return in getter expected to return 'T'}}
+  subscript(_ i: Int) -> Self {} // expected-error{{getter with uninhabited return type 'Never' is missing call to another never-returning function on all paths}}
+  subscript(_ s: Self) -> Self {}
+
+  static func unreachable_static_implicit_return(_ n: Never) -> Int {}
+  func unreachable(_ n: Never) -> Int { // expected-note{{'n' is of type 'Never' which cannot be constructed because it is an enum with no cases}}
+    42 // expected-warning{{will never be executed}}
+  }
+
+  // FIXME: should these produce unreachable code warnings since they implicity take an uninhabited 'self' param?
+  func implicitly_unreachable() { _ = 42 }
+  func implicitly_unreachable_implicit_return() -> Int { 42 }
+}
+
+enum InhabitedType {
+  case inhabitant
+
+  // Uninhabited params
+  subscript(_ n: Never) -> Int {}
+  subscript<T>(_ e: Never) -> T {}
+  subscript(_ v: MyEmptyType, e: Int) -> Never {}
+
+  // Inhabited params
+  subscript(_ i: Int) -> Int {} // expected-error{{missing return in getter expected to return 'Int'}}
+  subscript(_ j: Int) -> Void {}
+  subscript(_ k: Int) -> Never {} // expected-error{{getter with uninhabited return type 'Never' is missing call to another never-returning function on all paths}}
+  // FIXME: ^ this diagnostic should probably use the word 'subscript' rather than 'getter'
```

Comment:
noticed this wording change when moving to the SIL-based diagnostics. i have yet to determine how to special case the relevant logic to improve this. let me know if this seems like a blocker.

Reply:
I don't think it's a blocker, it's the same diagnostic we give today for e.g:

```swift
struct S {
  subscript(_ x: Int) -> Never {
    let x = 0
  }
}
```